### PR TITLE
Allow crontab to correctly handle daylight switching time

### LINKED ISSUE #10107: Hourly crontab skips execution during fall-back DST transition
## Description

Hourly crontab schedules skip one execution during fall-back DST transitions when the same local hour occurs twice.

During fall-back (e.g., when clocks go from 1:59 AM back to 1:00 AM), an hourly task that ran at the first 1:00 AM will not run at the second 1:00 AM, even though a full hour has elapsed in real time.

## Celery Version

- **Celery:** 5.6.2
- **Python:** 3.11+

## Reproduction

```python
from datetime import datetime
from zoneinfo import ZoneInfo
from celery.schedules import crontab

# Fall-back Nov 1, 2026 US/Pacific:
#   8:00 UTC = 1:00 AM PDT (before clocks fall back)
#   9:00 UTC = 1:00 AM PST (after clocks fall back)

schedule = crontab(minute='0', hour='*')
schedule.tz = ZoneInfo('US/Pacific')
schedule.nowfun = lambda: datetime(2026, 11, 1, 9, 0)  # 9 UTC = 1 AM PST
result = schedule.is_due(datetime(2026, 11, 1, 8, 0))  # 8 UTC = 1 AM PDT (last run)

print(f'is_due: {result.is_due}')  # False - BUG: should be True
```

**Expected:** `is_due=True` (one hour has passed in real time)  
**Actual:** `is_due=False` (task skips execution)

## Root Cause

In `crontab.remaining_estimate()`, when comparing `last_run_at` to the current time, the method doesn't account for the UTC offset change during fall-back transitions. Since both times map to the same local hour (1 AM), Celery incorrectly determines the task already ran.

## Impact

Any hourly scheduled task will miss one execution per year during the fall-back DST transition in timezones that observe DST. This affects:
- `0 * * * *` (top of hour)
- `5 * * * *` (any minute, every hour)
- Any `hour='*'` crontab schedule

## Suggested Fix

Detect when the UTC offset has decreased by 1 hour (fall-back) and adjust `last_run_at` accordingly before comparison:

```python
def remaining_estimate(self, last_run_at, ffwd=ffwd):
    def _adjust_time(t):
        return self.to_local(self.maybe_make_aware(t))

    utc_now = self.now()
    local_now = _adjust_time(utc_now)
    local_hour_ago = _adjust_time(utc_now - timedelta(hours=1))
    
    offset_now = local_now.utcoffset()
    offset_hour_ago = local_hour_ago.utcoffset()
    
    if offset_now is not None and offset_hour_ago is not None:
        time_fell_back = (offset_now - offset_hour_ago).total_seconds() == -3600.0
        cron_is_hourly = getattr(self, "_orig_hour", None) == "*"
        
        if time_fell_back and cron_is_hourly and (utc_now > last_run_at):
            last_run_at = last_run_at - timedelta(hours=1)
    
    return _old_remaining_estimate(self, last_run_at, ffwd)
```

## Related Issues

- #1604 - Original DST issue (closed 2016)
- #4403 - Partial DST fix (merged 2017) - did not address this hourly edge case

Happy to submit a PR with the fix and tests if this approach is acceptable.

### PR DESCRIPTION
**Pull request by Amli**
## Description
This pull request is an attempt to get the crontab handle daylight saving correctly:
- do not produce a datetime that does not exists (at DST start, when an hour disappears)
- do not produce a date an hour late (this currently happens at DST end, when an hour appears twice the same day)
- do not produce a next date in the past (remaining_estimate does that on dst end)

After [trying to hack](https://github.com/opendatasoft/celery/commit/c8c73a3a2e643ac80360dba16b85d094e6ec57ac) the current code, I ended up with a complete rewrite of the main function that computes the next run date for a crontab because the existing code was entirely ignoring timezone and makes it very difficult to just hack around it to get a correct result in all cases.

## Disclaimer
That's a relatively big change on the crontab code (even more for a first time contribution), I hesitated quite some time before opening this and I would understand if was too risky but, given the number of issues involving timezones and DST, I think it could be a good thing to have a code that handles timezone at its heart.
It is also highly possible I overlooked some cases (not aware datetime do not seem to be possible in this part of the code but I wonder), so I would be happy to get more insight.
I'm obviously available to discuss all this and ready to change the approach if needed.

## Changes
The new `crontab._next_occurence()` function produces a correct timezone aware date for the next expected crontab run.
It replaces the uses of `remaining()` and `remaining_delta()` (and the internal `_delta_to_next()`).

It does not work with delta at all, the delta is simply computed in `remaining_estimate()` as a simple difference between now and next_run_at (in utc).

The logic inside `_next_occurence()` is a mix of the existing `remaining_delta()`, for the part where we check the current day first to be as fast as possible and `_delta_to_next()` where we look for candidate days.

The code tries to be as fast as possible while only producing correct datetimes, so it is possibly a bit slower, but as far as I can tell not significantly slower.

The existing `remaining_delta()` was kept and modified to use the new code but still produce an `ffwd` instance.
AFAIK it is not used in the codebase except in an assert in the tests but I don't know if it considered to be a visible method or not.

The changes here do not try to modify the cron behaviour, if a crontab is scheduled at `2` on a day that skips that hour in this timezone, I will just not run, and the opposite if the `2` hour happens 2 times, it will run twice that day.

## Tests
I added a whole bunch of tests for cases around DST in multiple timezone to ensure my implementation was correct.
Some tests come from https://github.com/celery/celery/pull/10131 and in the end, there are probably too many tests and duplicated cases...

Fixes #10107

First time contribution, do not hesitate to tell me if things should be done differently.

### CODE REVIEW

**Comment by auvipy** [comment]
did you work on top of the existing open PR or you introduced complete new code?

**Comment by Amli** [comment]
That's complete new code, I started looking at and trying to adjust the changes made in https://github.com/celery/celery/pull/10131 only to find out there was cases that was not easily handled by a simple hack.

**Comment by auvipy** [comment]
@phanky1 @GangEunzzang I would highly appreciate your participation in reviewing this PR

**Comment by GangEunzzang** [comment]
I spent some time digging into this and testing the branch on macOS 15 and Linux (`python:3.12`) against `main` (`3e40f43`), including a few cases outside the new tests.

The timezone-aware approach looks like the right direction, and the new tests are useful — 69 of them fail on current `main`, so they're exercising real behavior differences rather than just adding coverage around the new implementation. The existing `test_schedules.py` / `test_beat.py` tests also pass on the branch (392 total).

I could reproduce #10107 as well. Simulating the Berlin fall-back minute by minute, `crontab(minute=0)` fires 7 times on this branch versus 6 on `main`, restoring the run in the repeated hour.

While testing a few other timezone/schedule combinations, I ran into some cases where the branch differs from `main`, including a couple where it raises instead of returning a date.

### 1. Sparse schedules can raise `RuntimeError` where `main` finds the next run

The 8-year search window in `_generate_run_days_after()` isn't enough for some valid schedules, and this isn't specific to DST:

```python
c = crontab(minute=0, hour=0, day_of_month=29, month_of_year=2)
c.tz = ZoneInfo("UTC")
c.nowfun = lambda: datetime(2096, 3, 1, tzinfo=c.tz)

c.remaining_estimate(datetime(2096, 3, 1, tzinfo=c.tz))

# main:   2104-02-29
# branch: RuntimeError: unable to find a next occurrence
```

`range(start.year, start.year + 8)` also excludes the eighth year, so 2096 → 2104 is out of reach either way.

I found similar cases with:

* `day_of_month=29, month_of_year=2, day_of_week=1` from 2024 (`main`: 2044)
* `day_of_month=10, month_of_year=9, day_of_week=4` from 2026 (`main`: 2037)

I also did a small sweep of 48 sparse schedules. The branch raised `RuntimeError` on 14 of them versus 12 on `main`; the two additional failures were leap-day + weekday combinations.

### 2. `_get_run_minute()` can raise `ValueError`

I can reproduce this with Lord Howe's 30-minute DST transition:

```python
c = crontab(minute=0)
c.tz = ZoneInfo("Australia/Lord_Howe")

# last run: 2026-10-04 01:00
# DST jumps from 02:00 -> 02:30

# ValueError: min() iterable argument is empty
```

`_get_run_minute()` does:

```python
min(m for m in self.minute if m >= d.minute)
```

After the 30-minute shift, `d.minute` can be 30 while `self.minute` is `{0}`, leaving the generator empty.

### 3. Some valid hours are skipped around DST transitions

I found a couple of cases where the branch skips a wall-clock time that still exists.

For `America/Santiago`, midnight doesn't exist on the DST-start day. With `crontab(minute=0, hour=1)`, `main` returns the valid 01:00 on that day, while the branch skips to the following day.

I also saw this with `Pacific/Chatham` (+12:45). Starting from 03:45 after the spring-forward transition, an hourly schedule returns 04:00 on `main`, but 05:00 on the branch.

I ran a 20,000-case differential check against `main` using random schedule specs/times across 10 timezones. There were 6 differences of this kind, all in Lord Howe and Chatham.

### 4. The next run can resolve to the same instant as `last_run_at`

In `Australia/Lord_Howe`, if `last_run_at` is the second 01:30 during the fall-back and the schedule is `crontab(minute=30)`, the branch can return that same instant (`2026-04-04T15:00Z`) as the next run instead of advancing to the following occurrence.

That also makes `is_due` true again for a run that has already happened.

### 5. Performance for common schedules

I also did a quick `timeit` comparison of `remaining_estimate()` only (200 calls × 5 repeats, median, same machine):

| schedule                          |  `main` |   branch |
| --------------------------------- | ------: | -------: |
| `minute=0, hour=9`                |  8.3 µs | 129.9 µs |
| `minute=0, hour=9, day_of_week=1` |  8.3 µs | 126.6 µs |
| `minute=0, hour=23`               |  7.7 µs |  89.3 µs |
| `minute=0`                        |  9.2 µs |  19.7 µs |
| `minute='*'`                      | 12.1 µs |  12.0 µs |
| `day_of_month=1, month_of_year=1` | 21.8 µs |   7.9 µs |

Sparse schedules actually get faster, but a normal daily schedule is roughly 15× slower. Beat calls this for every entry on each tick, so the cost scales with the number of schedules. I haven't measured the impact on overall Beat throughput.

### 6. Daily schedules during a repeated hour

One behavior I'm not sure about is how we want to treat explicitly scheduled hours during fall-back.

Using the same Berlin transition, `crontab(minute=0, hour=2)` fires at both 02:00 CEST and 02:00 CET on the branch, whereas `main` fires once.

For `hour='*'`, firing in both instances of the repeated hour is exactly what #10107 is asking for. For an explicitly daily 02:00 schedule, though, this changes the existing semantics.

Vixie cron distinguishes these cases and avoids running fixed-time jobs twice during a repeated interval — cronie's `src/cron.c` comments the fall-back branch as "the fixed-time jobs probably have already run, and should not be repeated" — so it may be worth deciding explicitly whether Celery wants the same behavior or whether firing on both folds is intentional.

Two small things I noticed while going through the changes:

* `_next_occurence` looks like a typo; I think this should be `_next_occurrence`.
* Since some of the tests originated in #10131, it may be worth preserving attribution to the original author with a `Co-authored-by:` trailer.

Overall, the timezone-aware approach looks promising, and it does fix the repeated-hour case from #10107. The main areas that seem worth another pass are the sparse-schedule search limit, non-hour DST transitions, and the intended behavior for fixed-time schedules during a repeated hour.

**Comment by Amli** [comment]
@GangEunzzang thanks a lot for the thorough review and checks !
There are definitely some things I overlooked there (I had "Australia/Lord_Howe" in mind, and forgot to check it but I was not aware of "Pacific/Chatham"...).

I'll do a first pass of fixes today.

File: celery/schedules.py [hunk 1]
```diff
@@ -578,6 +500,135 @@ def __setstate__(self, state: Mapping[str, Any]) -> None:
         super().__init__(**state)
         self._orig_kwargs = dict(state)
 
+    def _next_occurence(self, current: datetime) -> datetime:
```
Comment (auvipy) [inline celery/schedules.py]:
```suggestion
    def _next_occurrence(self, current: datetime) -> datetime:
```

**Review by auvipy — CHANGES_REQUESTED** [review]

**Comment by GangEunzzang** [comment]
@Amli Glad it helped!

**Comment by Amli** [comment]
Runtime and ValueError exceptions should be fixed.
The cases for timezones with an half hour dst offset (Lord Howe Islands being the only one I could find) and timezones with minute level UTC offest like `Pacific/Chatham` should be good now too. I will double check this tomorrow, the day has been long already.

I will look at what can be done regarding performance after that.

On another matter, do you have a preference on rebase vs merge for keeping the branch up to date ? I also need to include `Co-authored-by:` for code attribution but an empty commit could do if all is squashed in the end.

**Comment by Amli** [comment]
to verify the written tests are correct and for future reference, I ran the current tests with a simple implementation of `_next_occurrence` that cannot only be correct (adds 1 minute in UTC, check if local time matches the crontab, continue until match):

```python
    def _next_occurrence(self, current: datetime) -> datetime:
        candidate_utc = current.replace(second=0, microsecond=0).astimezone(timezone.utc)
        tz = current.tzinfo
        while range(26280000):  # 50 years * 365 days * 24 hours * 60 minutes
            candidate_utc = candidate_utc + timedelta(minutes=1)
            candidate_tz = candidate_utc.astimezone(tz)
            if (
                candidate_tz.month in self.month_of_year
                and candidate_tz.day in self.day_of_month
                and candidate_tz.isoweekday() % 7 in self.day_of_week
            ) and candidate_tz.hour in self.hour and candidate_tz.minute in self.minute:
                return candidate_tz

        raise RuntimeError(f"unable to find a next occurrence for crontab {self}, start from {current}")
```

**Comment by auvipy** [comment]
can you please allow maintainer to push option?

**Review by auvipy — CHANGES_REQUESTED** [review]
since this is a complete rewrite, I would also like to have enough integrations tests with the new unit tests as well... this will help to reduce subtle bugs

**Comment by Amli** [comment]
damn, I opened the fork from an organization account which [does not allow that](https://github.com/orgs/community/discussions/5634) 😬

Switching to the same branch on my personal account should work, trying this...

edit: obviously cannot switch PR origin on an PR opened form a fork, I would need to open a new one

**Comment by Amli** [comment]
@auvipy PR reopened https://github.com/celery/celery/pull/10608 and this time you should have the permission to push
(I will be AFK til monday starting more or less now)

**Comment by auvipy** [comment]
no problem thanks. as mentioned, this should be covered with proper integration tests as well

**Review by MehrazRumman — CHANGES_REQUESTED** [review]
### Bug: repeated hour missed in minute-offset zones
```python
tz = ZoneInfo("Pacific/Chatham")          # +13:45 -> +12:45, fall-back at 03:45
c = crontab(minute=45, hour=2); c.tz = tz
c._next_occurrence(datetime(2024, 4, 7, 2, 51, tzinfo=tz))
# got 2024-04-07 14:00Z (next day) / expected 2024-04-06 14:00Z (second pass)
```
Same on `main`, so not a regression — but it's the #10107 bug class this PR targets.

### Performance
`remaining_estimate()`, Berlin: `minute=0,hour=9` 11.9 µs → 85.7 µs (~7×).
Beat pays this per entry per tick. Cheap wins:
- `_has_run_hours_after()` + `_get_next_run_hour()` walk the hours twice —
  one `next(gen, None)` would halve it.
- `sorted(month_of_year/day_of_month)` on every call; precompute in `__init__`.

**Comment by auvipy** [comment]
closing this in favor of the other PR
